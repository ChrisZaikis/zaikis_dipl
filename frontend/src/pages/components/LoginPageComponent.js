import { Container, Row, Col, Form, Button, Alert } from "react-bootstrap";
import { useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import Spinner from "react-bootstrap/Spinner";
import "../../scss/button.scss";

const LoginPageComponent = ({
  loginUserApiRequest,
  reduxDispatch,
  setReduxUserState,
}) => {
  const [validated, setValidated] = useState(false);
  const [loginUserResponseState, setLoginUserResponseState] = useState({
    success: "",
    error: "",
    loading: false,
  });

  const navigate = useNavigate();

  const handleSubmit = (event) => {
    event.preventDefault();
    event.stopPropagation();
    const form = event.currentTarget.elements;

    const email = form.email.value;
    const password = form.password.value;
    const doNotLogout = form.doNotLogout.checked;

    if (event.currentTarget.checkValidity() === true && email && password) {
      setLoginUserResponseState({ loading: true });
      loginUserApiRequest(email, password, doNotLogout)
        .then((res) => {
          setLoginUserResponseState({
            success: res.success,
            loading: false,
            error: "",
          });

          if (res.userLoggedIn) {
            reduxDispatch(setReduxUserState(res.userLoggedIn));
          }

          if (res.success === "user logged in" && !res.userLoggedIn.isAdmin)
            window.location.assign("/");
          else window.location.assign("/admin/orders");
        })
        .catch((er) =>
          setLoginUserResponseState({
            error: er.response.data.message
              ? er.response.data.message
              : er.response.data,
          })
        );
    }

    setValidated(true);
  };

  return (
    <Container>
      <Row className="mt-5 justify-content-md-center">
        <Col md={6}>
          <h1>Σύνδεση</h1>
          <Form noValidate validated={validated} onSubmit={handleSubmit}>
            <Form.Group className="mb-3" controlId="formBasicEmail">
              <Form.Label>Διεύθυνση Email*</Form.Label>
              <Form.Control
                name="email"
                required
                type="email"
                placeholder=" email"
              />
            </Form.Group>
            <Form.Group className="mb-3" controlId="formBasicPassword">
              <Form.Label>Κωδικός Πρόσβασης*</Form.Label>
              <Form.Control
                name="password"
                required
                type="password"
                placeholder="Κωδικός Πρόσβασης"
              />
            </Form.Group>
            <Form.Group className="mb-3" controlId="formBasicCheckbox">
              <Form.Check
                name="doNotLogout"
                type="checkbox"
                label="Μην αποσυνδεθείτε"
              />
            </Form.Group>

            <Row className="pb-2">
              <Col>
                Δεν έχεις λογαριασμό?
                <Link to={"/register"}> Εγγραφή </Link>
              </Col>
            </Row>

            <button className="button-77" type="submit">
              {loginUserResponseState &&
              loginUserResponseState.loading === true ? (
                <Spinner
                  as="span"
                  animation="border"
                  size="sm"
                  role="status"
                  aria-hidden="true"
                />
              ) : (
                ""
              )}
              Σύνδεση
            </button>
            <Alert
              show={
                loginUserResponseState &&
                loginUserResponseState.error === "wrong credentials"
              }
              variant="danger"
            >
              Λάθος Διαπιστευτήρια
            </Alert>
          </Form>
        </Col>
      </Row>
    </Container>
  );
};

export default LoginPageComponent;
