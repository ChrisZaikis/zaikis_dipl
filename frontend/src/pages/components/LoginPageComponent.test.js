import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import LoginPageComponent from "./LoginPageComponent";
import { MemoryRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";

let loginUserApiRequest = () => {
  return new Promise((resolve, reject) => {
    resolve({
      success: "Ο χρήστης είναι συνδεδεμένος",
      userLoggedIn: {
        _id: "hd73hdhsj",
        name: "Χρήστος",
        lastName: "Ζαΐκης",
        email: "admin@admin.com",
        isAdmin: true,
        doNotLogout: false,
      },
    });
  });
};

let loginUserApiRequestError = () => {
  return new Promise((resolve, reject) => {
    reject({
      response: {
        data: {
          message: "wrong credentials",
        },
      },
    });
  });
};

const assign = window.location.assign;
beforeAll(() => {
  Object.defineProperty(window, "location", {
    value: { assign: jest.fn() },
  });
});
afterAll(() => {
  window.location.assign = assign;
});

test("if admin is logged in", async () => {
  render(
    <Router>
      <LoginPageComponent
        loginUserApiRequest={loginUserApiRequest}
        reduxDispatch={() => {}}
        setReduxUserState={() => {}}
      />
    </Router>
  );
  await waitFor(() => screen.getByLabelText("Email address"));
  expect(screen.getByLabelText("Email address")).toBeInTheDocument();
  expect(screen.getByLabelText("Password")).toBeInTheDocument();
  expect(screen.getByText("Login", { selector: "button" })).toBeInTheDocument();

  const emailField = screen.getByLabelText("Email address");
  const passwordField = screen.getByLabelText("Password");
  const submitButton = screen.getByText("Login", { selector: "button" });

  fireEvent.focus(emailField);
  fireEvent.change(emailField, { target: { value: "admin@admin.com" } });
  fireEvent.blur(emailField);
  fireEvent.focus(passwordField);
  fireEvent.change(passwordField, { target: { value: "admin@admin.com" } });
  fireEvent.blur(passwordField);

  fireEvent.click(submitButton);

  await waitFor(() => screen.getByLabelText("Email address"));
  expect(window.location.assign).toHaveBeenCalledWith("/admin/orders");
});
test("if wrong credentials", async () => {
  render(
    <Router>
      <LoginPageComponent
        loginUserApiRequest={loginUserApiRequestError}
        reduxDispatch={() => {}}
        setReduxUserState={() => {}}
      />
    </Router>
  );
  await waitFor(() => screen.getByLabelText("Email address"));
  expect(screen.getByLabelText("Email address")).toBeInTheDocument();
  expect(screen.getByLabelText("Password")).toBeInTheDocument();
  expect(screen.getByText("Login", { selector: "button" })).toBeInTheDocument();

  const emailField = screen.getByLabelText("Email address");
  const passwordField = screen.getByLabelText("Password");
  const submitButton = screen.getByText("Login", { selector: "button" });

  fireEvent.focus(emailField);
  fireEvent.change(emailField, { target: { value: "admin@admin.com" } });
  fireEvent.blur(emailField);
  fireEvent.focus(passwordField);
  fireEvent.change(passwordField, { target: { value: "admin@admin.com" } });
  fireEvent.blur(passwordField);

  fireEvent.click(submitButton);

  await waitFor(() => screen.getByText(/Wrong credentials/i));
  expect(screen.getByText(/Wrong credentials/i)).toBeInTheDocument();
});

test("create login snapshot", () => {
  const tree = renderer
    .create(
      <Router>
        <LoginPageComponent
          loginUserApiRequest={loginUserApiRequest}
          reduxDispatch={() => {}}
          setReduxUserState={() => {}}
        />
      </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
