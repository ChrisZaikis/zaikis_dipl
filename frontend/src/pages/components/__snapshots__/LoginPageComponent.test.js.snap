// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create login snapshot 1`] = `
<div
  className="container"
>
  <div
    className="mt-5 justify-content-md-center row"
  >
    <div
      className="col-md-6"
    >
      <h1>
        Login
      </h1>
      <form
        className=""
        noValidate={true}
        onSubmit={[Function]}
      >
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formBasicEmail"
          >
            Email address
          </label>
          <input
            className="form-control"
            id="formBasicEmail"
            name="email"
            placeholder="Enter email"
            required={true}
            type="email"
          />
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formBasicPassword"
          >
            Password
          </label>
          <input
            className="form-control"
            id="formBasicPassword"
            name="password"
            placeholder="Password"
            required={true}
            type="password"
          />
        </div>
        <div
          className="mb-3"
        >
          <div
            className="form-check"
          >
            <input
              className="form-check-input"
              disabled={false}
              id="formBasicCheckbox"
              name="doNotLogout"
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="formBasicCheckbox"
              title=""
            >
              Do not logout
            </label>
          </div>
        </div>
        <div
          className="pb-2 row"
        >
          <div
            className="col"
          >
            Don't you have an account?
            <a
              href="/register"
              onClick={[Function]}
            >
               Register 
            </a>
          </div>
        </div>
        <button
          className="btn btn-primary"
          disabled={false}
          type="submit"
        >
          
          Login
        </button>
      </form>
    </div>
  </div>
</div>
`;
